### PR TITLE
Updated remote error handling

### PR DESCRIPTION
**Pull request by Cari1111**
# Problem

@PacoSchatz and I noticed that the remote error handling that was introduced with a dummy handling function in #82 is not practical for debugging. The problem was that a disconnected player can cause remote errors, but they are not really that important because the player already disconnected, and everything is handled. But for connected players, the remote errors are important for debugging.
 
# Changes

This problem is trying to be solved here, by making sure that there are less remote errors and sorting the errors by connected and disconnected players.

## Action validation on client side

It is now checked that every action that is being sent is a list of floats. This makes sure that the client does not send something that would cause a remote error. But the actual action validation still happens on the server side. 

## Knowing which player is connected

Every player has now a variable `is_connected`. I'm not sure if that is the best solution. It would also be possible to do this with the connected player list in the player manager, but I thought this would be more complicated. But I am open for other suggestions here.

## Sending commands

Commands (e.g. to notify start, end, ...) are now only sent if the player is actually connected. This also reduces the number or remote errors.

## Handling remote errors

The remote errors are now handled by the `Server` by the method `on_remote_error()`. There are different error messages created for connected and disconnected players.

### CODE REVIEW

File: comprl/server/__main__.py [hunk 1]
```diff
@@ -64,6 +64,13 @@ def on_timeout(self, player: IPlayer, failure, timeout):
         log.debug(f"Player {player.id} had timeout after {timeout}s")
         player.disconnect(reason=f"Timeout after {timeout}s")
 
+    def on_remote_error(self, player: IPlayer, error: Exception):
+        """gets called when there is an error in deferred"""
+        if player.is_connected:
+            log.error(f"Connected player caused remote error \n {error.with_traceback}")
```
Comment (PacoSchatz) [inline comprl/server/__main__.py:70]: Maybe we can consider swapping `error.with_traceback` to just `error` as then the complete error and file/method it occurred is printed and not only a cryptical short version. As when an error with a logged-in player happens, this is interesting and might not be expected behavior.

I.e. when raising `raise ValueError("This is a test error")` in `_update()` in the game logic we get with  `error.with_traceback` just:
```ERROR:root:Connected player caused remote error 
 <built-in method with_traceback of Failure object at 0x00000273AA5260B0>
```

 But with  just `error`:
```
ERROR:root:Connected player caused remote error 
 [Failure instance: Traceback: <class 'ValueError'>: This is a test error
c:\Users\pacos\Documents\GitHub\teamproject-competition-server\.conda\Lib\site-packages\twisted\protocols\amp.py:1029:ampBoxReceived
c:\Users\pacos\Documents\GitHub\teamproject-competition-server\.conda\Lib\site-packages\twisted\protocols\amp.py:957:_answerReceived
c:\Users\pacos\Documents\GitHub\teamproject-competition-server\.conda\Lib\site-packages\twisted\internet\defer.py:874:callback     
c:\Users\pacos\Documents\GitHub\teamproject-competition-server\.conda\Lib\site-packages\twisted\internet\defer.py:981:_startRunCallbacks
--- <exception caught here> ---
c:\Users\pacos\Documents\GitHub\teamproject-competition-server\.conda\Lib\site-packages\twisted\internet\defer.py:1075:_runCallbacks
C:\Users\pacos\Documents\GitHub\teamproject-competition-server\comprl\server\networking.py:214:<lambda>
C:\Users\pacos\Documents\GitHub\teamproject-competition-server\comprl\server\interfaces.py:162:_res
C:\Users\pacos\Documents\GitHub\teamproject-competition-server\examples\hockey\hockey_game.py:64:_update
]
```
which is in my opinion more useful

**Review by PacoSchatz — CHANGES_REQUESTED** [review]
Thanks for your work. I tested it and it works as intended. That improves what errors are shown

Maybe we can consider printing every detail on errors with connected players, as these are the errors we are actually "interested in" most of the time. I marked it below

**Comment by Raining-Cloud** [comment]
Uhm I generally think its confusing that we have a connected flag in every player...
I don't see why there even is a state in which a non connected player causes problems and if there is such state this is an bug we should address and not just shortcust with this flag

File: comprl/server/__main__.py (hunk 1 above)
Comment (Cari1111) [reply comprl/server/__main__.py:70]: Thanks, for the suggestion. You are right, that makes more sense.

**Review by PacoSchatz — APPROVED** [review]
Thanks for your work. it works now as intended.

**Pull request merged**